### PR TITLE
8259984: IGV: Crash when drawing control flow before GCM

### PR DESCRIPTION
**Pull request by robcasloz**
The block formation algorithm applied by IGV to schedule graphs without associated CFG information traverses the graph backwards. This makes it difficult to deal with block projection nodes, leading in some cases to double addition of block nodes and block edges, and ultimately causing assertion failures.

This fix replaces the backward traversal by a forward traversal that relies on node category information (introduced in [8261336](https://bugs.openjdk.java.net/browse/JDK-8261336)) to identify control successors. The forward traversal is arguably simpler and, besides avoiding the reported assertion failure, has two advantages:

- it places block projection nodes in the same block as their predecessors, and
- it numbers basic blocks more naturally.

The following screenshots illustrate the improvements (before the fix to the left, after the fix to the right):

cfgs-before-after

Tested automatically on tens of thousands of graphs by running `java -Xcomp -XX:-TieredCompilation -XX:PrintIdealGraphLevel=4 ...` on an instrumented version of IGV that schedules graphs eagerly. Checked manually that the CFGs of a few selected graphs (included the reported one) are well-formed. Also checked that the overall IGV graph scheduling time is not affected by the changes.

Thanks to Christian Hagedorn for checking the fix independently.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8259984](https://bugs.openjdk.java.net/browse/JDK-8259984): IGV: Crash when drawing control flow before GCM

### Reviewers
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2607/head:pull/2607`
`$ git checkout pull/2607`

### CODE REVIEW

File: src/utils/IdealGraphVisualizer/ServerCompiler/src/com/sun/hotspot/igv/servercompiler/ServerCompilerScheduler.java [hunk 1]
```diff
@@ -620,5 +631,65 @@ public void buildUpGraph() {
                 toNode.preds.add(fromNode);
             }
         }
+
```
Comment (chhagedorn) [inline src/utils/IdealGraphVisualizer/ServerCompiler/src/com/sun/hotspot/igv/servercompiler/ServerCompilerScheduler.java:634]: Maybe you could extract the following code into 2 separate methods.

**Review by chhagedorn — APPROVED** [review]
Looks good to me! I played around with it in the IGV. I regularly got some assertion failure messages before - now they are gone now :+1:

**Review by neliasso — COMMENTED** [review]
Excellent!

File: src/utils/IdealGraphVisualizer/ServerCompiler/src/com/sun/hotspot/igv/servercompiler/ServerCompilerScheduler.java (hunk 1 above)
Comment (robcasloz) [reply src/utils/IdealGraphVisualizer/ServerCompiler/src/com/sun/hotspot/igv/servercompiler/ServerCompilerScheduler.java:637]: Thanks, done!

**Comment by robcasloz** [comment]
Thanks Christian and Nils for reviewing! Christian, I refactored the code as per your suggestion, please have a look again.

**Review by chhagedorn — APPROVED** [review]
Looks good!

**Review by neliasso — APPROVED** [review]

**Comment by robcasloz** [comment]
/summary
Replace backward traversal in the IGV block formation algorithm by forward
traversal guided by node category information. This change addresses the
reported assertion failures, places block projection nodes together with their
predecessors, and gives a more natural block numbering.

**Comment by robcasloz** [comment]
> Looks good!

Thanks again, Christian.

**Pull request closed without merging**